File: AoC2019/day2_prog_alarm/program_alarm/lib/program_alarm.ex
defmodule ProgramAlarm do

# comes in as array
# look at the first integer (position 0). Here, you will find an opcode - either 1, 2, or 99.
# 99 means program is finished and should halt. 
# Encountering an unknown opcode means something went wrong.
# Once you're done processing an opcode, move to the next one by stepping forward 4 positions.

  def build_alarm do
    data_set = prep_file_input

    fix = data_set
    # Step 1: Run opcode 1
    # Step 2: Step forward 4 positions
    # run an opcode
    # always steppign four positiosn forward

  end

  # Takes two positions
  def opcode_one(pos1, pos2) do
    # adds arr[pos1] + arr[pos2]
    # writes it in arr[pos3]

  end

  def opcode_two do
    
  end


  def prep_file_input do
    {:ok, file} = File.read("lib/input.txt")
    file 
      |> String.split(",")
      |> Enum.map(fn x -> String.to_integer(x) end) 
  end
  
end

File: AoC2019/python/day_2_program_alarm/program_alarm.py
import os

# 1) incoming Intcode program comes in as list of Integers
# 2) Look at position 0. Here, you find an opcode - either 1, 2, or 99.
#  - 99 means that  program is finished and should immediately halt.
#  - Opcode 1 adds together numbers read from two positions and stores the result in a third position. The three integers immediately after the opcode tell you these three positions - the first two indicate the positions from which you should read the input values, and the third indicates the position at which the output should be stored.

# For example, if your Intcode computer encounters 1,10,20,30, it should read the values at positions 10 and 20, add those values, and then overwrite the value at position 30 with their sum.

# Opcode 2 works exactly like opcode 1, except it multiplies the two inputs instead of adding them. Again, the three integers after the opcode indicate where the inputs and outputs are, not their values.

# Once you're done processing an opcode, move to the next one by stepping forward 4 positions.


class ProgramAlarm:

    PARENT_PATH = os.path.dirname(os.path.abspath(__file__))
    INTCODE_LOCATION = os.path.join(PARENT_PATH, "input.txt")

    def __init__(self, intcode_file=INTCODE_LOCATION):
        with open(intcode_file, 'r') as inputfile:
            self.intcode = inputfile.read()

# def read_program(self):
#     with open(os.path.join(os.path.dirname(os.path.abspath(__file__)), ), 'rt') as intcode:
#     pass


cp_alarm = ProgramAlarm()
